editor/src/clj/editor/gl/vertex2.clj

```diff
@@ -145,113 +96,89 @@
+                 :type-unsigned-int `(.putInt (.intValue (Long/valueOf (bit-and ~arg 0xffffffff))))))))
        ~'vbuf)))
 
-(defn- parse-attribute-definition
-  [form]
+(defn- parse-attribute-definition [form]
+  {:post [(graphics.types/attribute-info? %)]}
   (let [[type nm & [normalize]] form
         [prefix suffix]  (str/split (name type) #"\.")
         prefix           (keyword prefix)
         suffix           (keyword (or suffix "float"))
-        num-components   (case prefix
-                           :vec1 1
-                           :vec2 2
-                           :vec3 3
-                           :vec4 4)
+        vector-type      (case prefix
+                           :vec1 :vector-type-scalar
+                           :vec2 :vector-type-vec2
+                           :vec3 :vector-type-vec3
+                           :vec4 :vector-type-vec4)
+        data-type        (graphics.types/buffer-data-type-data-type suffix)
         attribute-name   (name nm)
-        attribute-key    (attribute-name->key attribute-name)
-        semantic-type    (attribute-key->semantic-type attribute-key)
-        vector-type      (case num-components
-                           1 :vector-type-scalar
-                           2 :vector-type-vec2
-                           3 :vector-type-vec3
-                           4 :vector-type-vec4)]
-    (assert num-components (str type " is not a valid type name. It must start with vec1, vec2, vec3, or vec4."))
-    (assert (get gl-types suffix) (str type " is not a valid type name. It must end with byte, short, int, float, or double. (Defaults to float if no suffix.)"))
+        attribute-key    (graphics.types/attribute-name-key attribute-name)
+        semantic-type    (graphics.types/infer-semantic-type attribute-key)
+
+        ;; TODO: You can't define a local-space attribute using defvertex.
+        coordinate-space (case semantic-type
+                           (:semantic-type-position :semantic-type-normal :semantic-type-tangent) :coordinate-space-world
+                           :coordinate-space-local)]
     {:name attribute-name
      :name-key attribute-key
-     :type suffix
-     :components num-components
-     :normalize (true? normalize)
-     :coordinate-space :coordinate-space-world
+     :normalize (boolean normalize)
+     :coordinate-space coordinate-space
+     :step-function :vertex-step-function-vertex
      :vector-type vector-type
+     :data-type data-type
      :semantic-type semantic-type}))
 
 (defmacro defvertex
   [name & attribute-definitions]
   (let [attributes         (mapv parse-attribute-definition attribute-definitions)
-        vertex-description (make-vertex-description name attributes)
+        vertex-description (-> attributes
+                               (graphics.types/make-vertex-description)
+                               (assoc :defvertex-name-sym name)) ;
```

Comment:
This used to be called `:name`. It only appears to be used by the tests at the moment.